test/common/duplexpair.js

```diff
@@ -24,8 +24,12 @@ class DuplexSocket extends Duplex {
   _write(chunk, encoding, callback) {
     assert.notStrictEqual(this[kOtherSide], null);
     assert.strictEqual(this[kOtherSide][kCallback], null);
-    this[kOtherSide][kCallback] = callback;
     this[kOtherSide].push(chunk);
+    if (!this[kOtherSide].readableLength) {
```

Comment:
If we have no readableLength _read will never be called.

Reply:
Just to be clear that I’m not misunderstanding – this is because _read() *was* already called?

Reply:
But wouldn’t `.push()` schedule another `._read()` call, because the readable side’s buffer is empty once the data from `.push()` is used?

Reply:
@addaleax `push(chunk)` should not schedule another read if `push(null)` is called in the same tick.

Reply:
The current implementation disregards this invariant of `Readable`.

Reply:
This made me consider why `push(null)` is called before `_read`. Could be that `_final` is called before all writes having finished. Let me check.

Reply:
no, it simply that an empty push will not trigger another read, as it is sensible to assume that if we couldn't push data then there is no data

Reply:
I mean, that kind of makes sense, but it’s also weird that `.push()` with an empty chunk leads to different behaviour than `.push()` with a non-empty chunk…

Reply:
Yea, I've been trying to get my head around it for a few days. But as you write, it's weird but makes sense. The alternative is more confusing and leads to other problems such as livelocks.